File: app/src/main/java/com/imagegallery/list/ImageAdapter.java
package com.imagegallery.list;

import android.content.Context;
import android.graphics.Point;
import android.support.v7.widget.RecyclerView;
import android.view.Display;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;

import com.imagegallery.R;
import com.imagegallery.model.PhotoSearchResultItem;

import java.util.List;

class ImageAdapter extends RecyclerView.Adapter<ImageViewHolder> {

    private final int columns;
    private final List<PhotoSearchResultItem> images;

    private OnPhotoSearchResultClickListener searchResultClickListener;

    interface OnPhotoSearchResultClickListener {
        void onImageClicked(PhotoSearchResultItem image);
    }

    ImageAdapter(List<PhotoSearchResultItem> images, int columns) {
        this.images = images;
        this.columns = columns;
    }

    @Override
    public ImageViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = LayoutInflater.from(context);

        View view = inflater.inflate(R.layout.layout_list_image, parent, false);

        WindowManager wm = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        Display display = wm.getDefaultDisplay();
        Point size = new Point();
        display.getSize(size);

        return new ImageViewHolder(view, searchResultClickListener, size.x / columns);
    }

    @Override
    public void onBindViewHolder(ImageViewHolder holder, int position) {
        holder.setImage(images.get(position));
    }

    @Override
    public int getItemCount() {
        return images.size();
    }

    void setOnImageClickListener(OnPhotoSearchResultClickListener imageClickListener) {
        this.searchResultClickListener = imageClickListener;
    }
}

File: app/src/main/java/com/imagegallery/list/ImageAdapter.kt
package com.imagegallery.list

import android.content.Context
import android.graphics.Point
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import android.view.WindowManager
import com.imagegallery.R
import com.imagegallery.model.PhotoSearchResultItem

internal class ImageAdapter(private val images: List<PhotoSearchResultItem>, private val columns: Int) : RecyclerView.Adapter<ImageViewHolder>() {

    private var searchResultClickListener: OnPhotoSearchResultClickListener? = null

    internal interface OnPhotoSearchResultClickListener {
        fun onImageClicked(image: PhotoSearchResultItem)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ImageViewHolder {
        val context = parent.context
        val inflater = LayoutInflater.from(context)

        val view = inflater.inflate(R.layout.layout_list_image, parent, false)

        val wm = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val display = wm.defaultDisplay
        val size = Point()
        display.getSize(size)

        return ImageViewHolder(view, searchResultClickListener!!, size.x / columns)
    }

    override fun onBindViewHolder(holder: ImageViewHolder, position: Int) {
        holder.setImage(images[position])
    }

    override fun getItemCount(): Int {
        return images.size
    }

    fun setOnImageClickListener(imageClickListener: OnPhotoSearchResultClickListener) {
        this.searchResultClickListener = imageClickListener
    }
}